tasks/remote_storage.py

```diff
@@ -1,9 +1,221 @@
+Jinja2 template and putting it into every directory. Then we upload particular
+job directory using awscli (which does parallelism for us) under S3 bucket
+"jobs" prefix (directory). Currently 2 awscli commands are needed in order to
+set correct encoding for gzip files. Content types are set automatically using
+"/etc/mime.types" file.
+At the end we create root jobs index to list all "freeipa" repo PR jobs in the
+bucket.
+"""
+
+
+def get_jobdir_metadata(job_uuid):
+    """
+    Job metadata for root index.
+    """
+    try:
+        client = boto3.client('s3')
+        job = client.head_object(Bucket=CLOUD_BUCKET, Key=job_uuid)
+        name = job_uuid.split(os.sep)[-2]
+        if re.match(UUID_RE, name):
+            mtime = job.get('LastModified')
+            metadata = job.get('Metadata', dict())
+            repo_owner = metadata.get('repo_owner')
+            pr_number = metadata.get('pr_number')
+            pr_author = metadata.get('pr_author')
+            task_name = metadata.get('task_name')
+            returncode = metadata.get('returncode')
+            if repo_owner == 'freeipa':
+                return {'name': name,
+                        'mtime': mtime,
+                        'pr_number': pr_number,
+                        'pr_author': pr_author,
+                        'task_name': task_name,
+                        'returncode': returncode}
+    except ClientError:
+        # lets not fail in case we cannot read metadata of some directory or
+        # there appears a rogue directory without metadata (returns None then)
+        pass
+
+
+def get_jobs():
+    """
+    Get all jobs (directories/objects) from root "jobs" directory. Pagination
+    is needed as S3 client returns only 1000 objects at once.
+    """
+    try:
+        client = boto3.client('s3')
+        paginator = client.get_paginator('list_objects_v2')
+        iterator = paginator.paginate(Bucket=CLOUD_BUCKET,
+                                      Prefix=CLOUD_JOBS_DIR,
+                                      Delimiter='/',
+                                      PaginationConfig={'PageSize': None})
+        for job_uuid in iterator.search('CommonPrefixes'):
+
+            job_dir = job_uuid['Prefix']
+            yield job_dir
+    except ClientError as exc:
+        logging.error('Error while getting job directory list')
+        logging.debug(exc, exc_info=True)
+
+
+def create_jobs_root_index():
+    """
+    Generate root jobs index (only directories) using boto3 client. Threading
+    is needed as polling S3 objects serially is very slow.
+    """
+    try:
+        client = boto3.client('s3')
+        objects = []
+        with ThreadPoolExecutor(max_workers=100) as executor:
+            todo = []
+            for job in get_jobs():
+                future = executor.submit(get_jobdir_metadata, job)
+                todo.append(future)
+            for future in concurrent.futures.as_completed(todo):
+                res = future.result()
```

Comment:
Here you can check iff `res` is not `None` to append it to objects, with that you will get rid of line 101